torchrl/data/utils.py

```diff
@@ -35,6 +43,206 @@
+                            if re.match(
+                                r"Found more than one unique shape in the tensors",
+                                str(err),
+                            ):
+                                # sub_td_key is het leaf so lets recurse to a dense version of it to get the example
+                                temp_td = sub_td
+                                while hasattr(
+                                    temp_td, "tensordicts"
+                                ):  # we need to grab the het tensor from the inner nesting level
+                                    temp_td = sub_td.tensordicts[0]
+                                value = temp_td.get(sub_td_key)
+                            else:
+                                raise err
+                        lazy_keys_examples.update({sub_td_key: value})
+
+        for td_index in range(len(td.tensordicts)):  # add missing lazy entries
+            sub_td = td.tensordicts[td_index]
+            for lazy_key in set(lazy_keys_examples.keys()).difference(
+                lazy_keys_per_td[td_index]
+            ):
+                lazy_key_example = lazy_keys_examples[lazy_key]
+                sub_td.set(
+                    lazy_key,
+                    _empty_like(lazy_key_example, sub_td.batch_size),
+                )
+            td.tensordicts[td_index] = sub_td
+
+    if recurse_through_entries:
+        for key in td.keys():
+            try:
+                td.set(
+                    key,
+                    unlazyfy_keys(
+                        td.get(key), recurse_through_entries, recurse_through_stack
+                    ),
+                )
+            except RuntimeError as err:
+                if re.match(
+                    r"Found more than one unique shape in the tensors", str(err)
+                ):
+                    pass  # no need to unlazify since key is het leaf
+                else:
+                    raise err
+
+    return td
+
+
+def relazyfy_keys(
+    td,
+):
+    """Given a TensorDictBase, restores lazy keys by removing 0 shaped tensors and related orphan tensordicts."""
+    if not is_tensor_collection(td):
+        return None if td.numel() == 0 else td.clone()
+
+    td = td.clone()
+
+    if isinstance(td, LazyStackedTensorDict):
+        for td_index in range(len(td.tensordicts)):
+            sub_td = td.tensordicts[td_index]
+            sub_td = relazyfy_keys(sub_td)
+            td.tensordicts[td_index] = sub_td
+
+    for key in list(td.keys()):
+        try:
+            value = relazyfy_keys(td.get(key))
+            if value is None:
+                del td[key]
+            else:
+                td.set(
+                    key,
+                    value,
+                )
+        except RuntimeError as err:
+            if re.match(r"Found more than one unique shape in the tensors", str(err)):
+                pass
+            else:
+                raise err
```

Comment:
capturing errors at runtime is costly, if this is expected to run once in a while we should find another way of doing it

Reply:
solved it, we might consider updating this too https://github.com/pytorch-labs/tensordict/blob/main/tensordict/tensordict.py#L8009 it is run at every print